File: app/src/androidTest/java/de/westnordost/streetcomplete/quests/AddCyclewayTest.java
package de.westnordost.streetcomplete.quests;

import de.westnordost.streetcomplete.data.osm.OsmElementQuestType;
import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
import de.westnordost.streetcomplete.data.osm.changes.StringMapEntryAdd;
import de.westnordost.streetcomplete.quests.bikeway.AddCycleway;
import de.westnordost.streetcomplete.quests.bikeway.AddCyclewayForm;
import de.westnordost.streetcomplete.quests.bikeway.Cycleway;

public class AddCyclewayTest extends AOsmElementQuestTypeTest
{
	public void testCyclewayLeftAndRightDontHaveToBeSpecified1()
	{
		bundle.putString(AddCyclewayForm.CYCLEWAY_LEFT, Cycleway.EXCLUSIVE_LANE.name());
		StringMapChangesBuilder cb = new StringMapChangesBuilder(tags);
		createQuestType().applyAnswerTo(bundle, cb);
		// success if no exception thrown
	}

	public void testCyclewayLeftAndRightDontHaveToBeSpecified2()
	{
		bundle.putString(AddCyclewayForm.CYCLEWAY_RIGHT, Cycleway.EXCLUSIVE_LANE.name());
		StringMapChangesBuilder cb = new StringMapChangesBuilder(tags);
		createQuestType().applyAnswerTo(bundle, cb);
		// success if no exception thrown
	}

	public void testCyclewayLane()
	{
		putBothSides(Cycleway.EXCLUSIVE_LANE);
		verify(
			new StringMapEntryAdd("cycleway:both", "lane"),
			new StringMapEntryAdd("cycleway:both:lane", "exclusive"));
	}

	public void testCyclewayAdvisoryLane()
	{
		putBothSides(Cycleway.ADVISORY_LANE);
		verify(
			new StringMapEntryAdd("cycleway:both", "lane"),
			new StringMapEntryAdd("cycleway:both:lane", "advisory"));
	}

	public void testCyclewayUnspecifiedLane()
	{
		putBothSides(Cycleway.LANE_UNSPECIFIED);
		verify(new StringMapEntryAdd("cycleway:both", "lane"));
	}

	public void testCyclewayTrack()
	{
		putBothSides(Cycleway.TRACK);
		verify(new StringMapEntryAdd("cycleway:both", "track"));
	}

	public void testCyclewayBusLane()
	{
		putBothSides(Cycleway.BUSWAY);
		verify(new StringMapEntryAdd("cycleway:both", "share_busway"));
	}

	public void testCyclewayPictogramLane()
	{
		putBothSides(Cycleway.PICTOGRAMS);
		verify(
			new StringMapEntryAdd("cycleway:both", "shared_lane"),
			new StringMapEntryAdd("cycleway:both:lane", "pictogram")
		);
	}

	public void testCyclewaySuggestionLane()
	{
		putBothSides(Cycleway.SUGGESTION_LANE);
		verify(
			new StringMapEntryAdd("cycleway:both", "shared_lane"),
			new StringMapEntryAdd("cycleway:both:lane", "advisory")
		);
	}

	public void testCyclewayNone()
	{
		putBothSides(Cycleway.NONE);
		verify(new StringMapEntryAdd("cycleway:both", "no"));
	}

	public void testCyclewayOnSidewalk()
	{
		putBothSides(Cycleway.SIDEWALK_EXPLICIT);
		verify(
				new StringMapEntryAdd("cycleway:both", "track"),
				new StringMapEntryAdd("sidewalk", "both"),
				new StringMapEntryAdd("cycleway:both:segregated", "no")
		);
	}

	public void testCyclewaySidewalkOkay()
	{
		putBothSides(Cycleway.SIDEWALK_OK);
		verify(
				new StringMapEntryAdd("cycleway:both", "no"),
				new StringMapEntryAdd("sidewalk", "both"),
				new StringMapEntryAdd("sidewalk:both:bicycle", "yes")
		);
	}

	public void testCyclewaySidewalkAny()
	{
		bundle.putString(AddCyclewayForm.CYCLEWAY_RIGHT, Cycleway.SIDEWALK_EXPLICIT.name());
		bundle.putString(AddCyclewayForm.CYCLEWAY_LEFT, Cycleway.SIDEWALK_OK.name());
		verify(
				new StringMapEntryAdd("sidewalk", "both")
		);
	}

	public void testCyclewayDualTrack()
	{
		putBothSides(Cycleway.DUAL_TRACK);
		verify(
			new StringMapEntryAdd("cycleway:both", "track"),
			new StringMapEntryAdd("cycleway:both:oneway", "no")
		);
	}

	public void testCyclewayDualLane()
	{
		putBothSides(Cycleway.DUAL_LANE);
		verify(
				new StringMapEntryAdd("cycleway:both", "lane"),
				new StringMapEntryAdd("cycleway:both:oneway", "no")
		);
	}

	public void testLeftAndRightAreDifferent()
	{
		bundle.putString(AddCyclewayForm.CYCLEWAY_RIGHT, Cycleway.EXCLUSIVE_LANE.name());
		bundle.putString(AddCyclewayForm.CYCLEWAY_LEFT, Cycleway.TRACK.name());
		verify(
				new StringMapEntryAdd("cycleway:right", "lane"),
				new StringMapEntryAdd("cycleway:right:lane","exclusive"),
				new StringMapEntryAdd("cycleway:left", "track")
		);
	}

	public void testCyclewayMakesStreetNotOnewayForBicycles()
	{
		putBothSides(Cycleway.EXCLUSIVE_LANE);
		bundle.putBoolean(AddCyclewayForm.IS_ONEWAY_NOT_FOR_CYCLISTS, true);
		verify(
				new StringMapEntryAdd("cycleway:both", "lane"),
				new StringMapEntryAdd("oneway:bicycle", "no"),
				new StringMapEntryAdd("cycleway:both:lane", "exclusive")
		);
	}

	public void testCyclewayLaneWithExplicitDirection()
	{
		// this would be a street that has lanes on both sides but is oneway=yes (in countries with
		// right hand traffic)
		putBothSides(Cycleway.EXCLUSIVE_LANE);
		bundle.putInt(AddCyclewayForm.CYCLEWAY_LEFT_DIR, -1);
		verify(
				new StringMapEntryAdd("cycleway:left", "lane"),
				new StringMapEntryAdd("cycleway:left:oneway", "-1"),
				new StringMapEntryAdd("cycleway:right", "lane"),
				new StringMapEntryAdd("cycleway:left:lane","exclusive"),
				new StringMapEntryAdd("cycleway:right:lane","exclusive")
		);
	}

	public void testCyclewayLaneWithExplicitOtherDirection()
	{
		// this would be a street that has lanes on both sides but is oneway=-1 (in countries with
		// right hand traffic)
		putBothSides(Cycleway.EXCLUSIVE_LANE);
		bundle.putInt(AddCyclewayForm.CYCLEWAY_LEFT_DIR, +1);
		verify(
				new StringMapEntryAdd("cycleway:left", "lane"),
				new StringMapEntryAdd("cycleway:left:oneway", "yes"),
				new StringMapEntryAdd("cycleway:right", "lane"),
				new StringMapEntryAdd("cycleway:left:lane","exclusive"),
				new StringMapEntryAdd("cycleway:right:lane","exclusive")
		);
	}

	private void putBothSides(Cycleway cycleway)
	{
		bundle.putString(AddCyclewayForm.CYCLEWAY_RIGHT, cycleway.name());
		bundle.putString(AddCyclewayForm.CYCLEWAY_LEFT, cycleway.name());
	}

	@Override protected OsmElementQuestType createQuestType()
	{
		return new AddCycleway(null);
	}
}

File: app/src/androidTest/java/de/westnordost/streetcomplete/quests/AddCyclewayTest.kt
package de.westnordost.streetcomplete.quests

import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder
import de.westnordost.streetcomplete.data.osm.changes.StringMapEntryAdd
import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao
import de.westnordost.streetcomplete.quests.bikeway.AddCycleway
import de.westnordost.streetcomplete.quests.bikeway.AddCyclewayForm
import de.westnordost.streetcomplete.quests.bikeway.Cycleway
import org.junit.Test

import org.mockito.Mockito.mock

class AddCyclewayTest : AOsmElementQuestTypeTest() {

    override val questType = AddCycleway(mock(OverpassMapDataDao::class.java))

    @Test fun cyclewayLeftAndRightDontHaveToBeSpecified1() {
        bundle.putString(AddCyclewayForm.CYCLEWAY_LEFT, Cycleway.EXCLUSIVE_LANE.name)
        val cb = StringMapChangesBuilder(tags)
        questType.applyAnswerTo(bundle, cb)
        // success if no exception thrown
    }

    @Test fun cyclewayLeftAndRightDontHaveToBeSpecified2() {
        bundle.putString(AddCyclewayForm.CYCLEWAY_RIGHT, Cycleway.EXCLUSIVE_LANE.name)
        val cb = StringMapChangesBuilder(tags)
        questType.applyAnswerTo(bundle, cb)
        // success if no exception thrown
    }

    @Test fun cyclewayLane() {
        putBothSides(Cycleway.EXCLUSIVE_LANE)
        verify(
            StringMapEntryAdd("cycleway:both", "lane"),
            StringMapEntryAdd("cycleway:both:lane", "exclusive")
        )
    }

    @Test fun cyclewayAdvisoryLane() {
        putBothSides(Cycleway.ADVISORY_LANE)
        verify(
            StringMapEntryAdd("cycleway:both", "lane"),
            StringMapEntryAdd("cycleway:both:lane", "advisory")
        )
    }

    @Test fun cyclewayUnspecifiedLane() {
        putBothSides(Cycleway.LANE_UNSPECIFIED)
        verify(StringMapEntryAdd("cycleway:both", "lane"))
    }

    @Test fun cyclewayTrack() {
        putBothSides(Cycleway.TRACK)
        verify(StringMapEntryAdd("cycleway:both", "track"))
    }

    @Test fun cyclewayBusLane() {
        putBothSides(Cycleway.BUSWAY)
        verify(StringMapEntryAdd("cycleway:both", "share_busway"))
    }

    @Test fun cyclewayPictogramLane() {
        putBothSides(Cycleway.PICTOGRAMS)
        verify(
            StringMapEntryAdd("cycleway:both", "shared_lane"),
            StringMapEntryAdd("cycleway:both:lane", "pictogram")
        )
    }

    @Test fun cyclewaySuggestionLane() {
        putBothSides(Cycleway.SUGGESTION_LANE)
        verify(
            StringMapEntryAdd("cycleway:both", "shared_lane"),
            StringMapEntryAdd("cycleway:both:lane", "advisory")
        )
    }

    @Test fun cyclewayNone() {
        putBothSides(Cycleway.NONE)
        verify(StringMapEntryAdd("cycleway:both", "no"))
    }

    @Test fun cyclewayOnSidewalk() {
        putBothSides(Cycleway.SIDEWALK_EXPLICIT)
        verify(
            StringMapEntryAdd("cycleway:both", "track"),
            StringMapEntryAdd("sidewalk", "both"),
            StringMapEntryAdd("cycleway:both:segregated", "no")
        )
    }

    @Test fun cyclewaySidewalkOkay() {
        putBothSides(Cycleway.SIDEWALK_OK)
        verify(
            StringMapEntryAdd("cycleway:both", "no"),
            StringMapEntryAdd("sidewalk", "both"),
            StringMapEntryAdd("sidewalk:both:bicycle", "yes")
        )
    }

    @Test fun cyclewaySidewalkAny() {
        bundle.putString(AddCyclewayForm.CYCLEWAY_RIGHT, Cycleway.SIDEWALK_EXPLICIT.name)
        bundle.putString(AddCyclewayForm.CYCLEWAY_LEFT, Cycleway.SIDEWALK_OK.name)
        verify(
            StringMapEntryAdd("sidewalk", "both")
        )
    }

    @Test fun cyclewayDualTrack() {
        putBothSides(Cycleway.DUAL_TRACK)
        verify(
            StringMapEntryAdd("cycleway:both", "track"),
            StringMapEntryAdd("cycleway:both:oneway", "no")
        )
    }

    @Test fun cyclewayDualLane() {
        putBothSides(Cycleway.DUAL_LANE)
        verify(
            StringMapEntryAdd("cycleway:both", "lane"),
            StringMapEntryAdd("cycleway:both:oneway", "no")
        )
    }

    @Test fun leftAndRightAreDifferent() {
        bundle.putString(AddCyclewayForm.CYCLEWAY_RIGHT, Cycleway.EXCLUSIVE_LANE.name)
        bundle.putString(AddCyclewayForm.CYCLEWAY_LEFT, Cycleway.TRACK.name)
        verify(
            StringMapEntryAdd("cycleway:right", "lane"),
            StringMapEntryAdd("cycleway:right:lane", "exclusive"),
            StringMapEntryAdd("cycleway:left", "track")
        )
    }

    @Test fun cyclewayMakesStreetNotOnewayForBicycles() {
        putBothSides(Cycleway.EXCLUSIVE_LANE)
        bundle.putBoolean(AddCyclewayForm.IS_ONEWAY_NOT_FOR_CYCLISTS, true)
        verify(
            StringMapEntryAdd("cycleway:both", "lane"),
            StringMapEntryAdd("oneway:bicycle", "no"),
            StringMapEntryAdd("cycleway:both:lane", "exclusive")
        )
    }

    @Test fun cyclewayLaneWithExplicitDirection() {
        // this would be a street that has lanes on both sides but is oneway=yes (in countries with
        // right hand traffic)
        putBothSides(Cycleway.EXCLUSIVE_LANE)
        bundle.putInt(AddCyclewayForm.CYCLEWAY_LEFT_DIR, -1)
        verify(
            StringMapEntryAdd("cycleway:left", "lane"),
            StringMapEntryAdd("cycleway:left:oneway", "-1"),
            StringMapEntryAdd("cycleway:right", "lane"),
            StringMapEntryAdd("cycleway:left:lane", "exclusive"),
            StringMapEntryAdd("cycleway:right:lane", "exclusive")
        )
    }

    @Test fun cyclewayLaneWithExplicitOtherDirection() {
        // this would be a street that has lanes on both sides but is oneway=-1 (in countries with
        // right hand traffic)
        putBothSides(Cycleway.EXCLUSIVE_LANE)
        bundle.putInt(AddCyclewayForm.CYCLEWAY_LEFT_DIR, +1)
        verify(
            StringMapEntryAdd("cycleway:left", "lane"),
            StringMapEntryAdd("cycleway:left:oneway", "yes"),
            StringMapEntryAdd("cycleway:right", "lane"),
            StringMapEntryAdd("cycleway:left:lane", "exclusive"),
            StringMapEntryAdd("cycleway:right:lane", "exclusive")
        )
    }

    private fun putBothSides(cycleway: Cycleway) {
        bundle.putString(AddCyclewayForm.CYCLEWAY_RIGHT, cycleway.name)
        bundle.putString(AddCyclewayForm.CYCLEWAY_LEFT, cycleway.name)
    }
}
